### ISSUE TITLE
Add a MTLFence wait on timestamp MTLBlitCommandEncoders

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/MoltenVK#1433]
Based on code review feedback from Apple, when using `MVKCommandEncoder::encodeTimestampStageCounterSamples()`, add a `MTLFence` between the  operational `MTLCommandEncoder` and the `MTLBlitCommandEncoders` used to record timestamps, to ensure the timestamp waits for existing work to finish (since there are no obvious task dependencies that Metal can detect in this situation).

**Comment by tomgreen66** [comment]
Would this produce (tried tracing how this is reached and looks like a timestamp creation so thought it might be related):
```
-[MTLDebugBlitCommandEncoder sampleCountersInBuffer:atSampleIndex:withBarrier:]:1682: failed assertion `sampleIndex(0) must be < sampleCount(0).'
```

Currently seeing this when compiling debug version of RBDOOM-3-BFG.  Big Sur - late 2014 MacBook with Intel Iris Pro GPU.  Currently having Vulkan issue with skinned models not appearing in the game which looks like an Intel driver issue since all other tests on other platforms report no problem. Just checking latest Vulkan SDK for any differences.

**Comment by billhollings** [comment]
> Would this produce
-[MTLDebugBlitCommandEncoder sampleCountersInBuffer:atSampleIndex:withBarrier:]:1682: failed assertion `sampleIndex(0) must be < sampleCount(0).'

Highly unlikely. 

`MVKCommandEncoder::encodeTimestampStageCounterSamples()` runs on Apple Silicon only, whereas the Metal call you are erroring on only runs on IM GPUs (Intel included).

However...the error you are encountering is a problem. The basic issue is that the app seems to be trying to write a timestamp to a query in a query pool that has `VkQueryPoolCreateInfo::queryCount` set to zero. Is it possible for you to verify that?

However, it might also be possible that the use of older hardware is triggering something obscure in the setup of the `MTLCounterSet` in Metal.

Are you building MoltenVK yourself? If so, what version of Xcode are you building MoltenVK with?

**Comment by tomgreen66** [comment]
Just downloaded latest version on LunarG website (1.2.189.0) and build RBDOOM-3-BFG via Xcode. 1.2.182.0 worked fine.  The backtrace of the failure:
```
    frame #6: 0x00007fff20b5aae0 MetalTools`-[MTLDebugBlitCommandEncoder sampleCountersInBuffer:atSampleIndex:withBarrier:] + 358
    frame #7: 0x0000000118d7df2d libMoltenVK.dylib`MVKCommandEncoder::markTimestamp(MVKTimestampQueryPool*, unsigned int) + 205
    frame #8: 0x0000000118d4ce31 libMoltenVK.dylib`MVKTimestampQueryPool::endQuery(unsigned int, MVKCommandEncoder*) + 33
    frame #9: 0x0000000118d7af8b libMoltenVK.dylib`MVKCommandEncoder::encode(id<MTLCommandBuffer>, MVKCommandEncodingContext*) + 155
    frame #10: 0x0000000118d7ae13 libMoltenVK.dylib`MVKCommandBuffer::submit(MVKQueueCommandBufferSubmission*, MVKCommandEncodingContext*) + 227
    frame #11: 0x0000000118dc72b4 libMoltenVK.dylib`MVKQueueFullCommandBufferSubmission<1ul>::submitCommandBuffers() + 68
    frame #12: 0x0000000118dc547b libMoltenVK.dylib`MVKQueueCommandBufferSubmission::execute() + 203
    frame #13: 0x0000000118dc41c4 libMoltenVK.dylib`MVKQueue::submit(unsigned int, VkSubmitInfo const*, VkFence_T*, MVKCommandUse) + 292
    frame #14: 0x0000000118d320fd libMoltenVK.dylib`vkQueueSubmit + 285
  * frame #15: 0x0000000100758ff2 RBDoom3BFG`idRenderBackend::GL_EndFrame(this=0x0000000101aedeb8) at RenderBackend_VK.cpp:2143:2
 ```

I am not building MoltenVK myself but if you need any further information I am happy to provide it.  Rather strangely at debug level the code itself doesn't include MoltenVK so assume the ICD is picking this up via my sourcing of `setup-env.sh` in the VulkanSDK/<version> directory.  I repeat for previous version 1.2.182.0 and it works so seems to be a change in behaviour.

**Comment by SRSaunders** [comment]
Not sure if this will help, but RBDOOM-3-BFG running with Vulkan SDK 1.2.189.0 works fine on a relatively modern iMac 2017 with AMD RX580 graphics running macOS Big Sur.  MoltenVK's new Metal timestamp support is very cool and is utilized by the game's detailed FPS counter.

However, I can duplicate the failure seen above when using SDK 1.2.189.0 on an older MacBook Pro 2012 with Intel HD 4000 graphics when when running RBDOOM-3-BFG on macOS Catalina.  The identical error occurs: `failed assertion sampleIndex(0) must be < sampleCount(0)`.  Previous Vulkan SDKs do not cause this failure - however I suspect the issue may be related to MoltenVK's new feature set so this probably makes sense.

Interestingly, when running the same MacBook Pro 2012 + SDK 1.2.189.0 setup on Mojave, the failure does not occur but it appears the timestamp feature is not showing any data either.

**Comment by billhollings** [comment]
Both the BLIT encoder sync AND the other issue:

>[MTLDebugBlitCommandEncoder sampleCountersInBuffer:atSampleIndex:withBarrier:]:1682: failed assertion `sampleIndex(0) must be < sampleCount(0).'

should be fixed with this PR.

Please retest.

**Comment by tomgreen66** [comment]
Just ran with compiled MoltenVK from head and it seems to work and not crash on startup.  My Intel Iris Pro is obviously showing it age.

**Comment by billhollings** [comment]
@tomgreen66 

> Just ran with compiled MoltenVK from head and it seems to work and not crash on startup. My Intel Iris Pro is obviously showing it age.

Thanks. I'm going to close this issue now.

@SRSaunders  

>I can duplicate the failure seen above when using SDK 1.2.189.0 on an older MacBook Pro 2012 with Intel HD 4000 graphics when when running RBDOOM-3-BFG on macOS Catalina

If you perform further testing with a build from the latest MoltenVK, and this issue has not been resolved in your environment, please reopen this issue and provide additional details.

**Comment by SRSaunders** [comment]
Verified that fix disables Metal timestamps on Intel HD 4000 (running Catalina) and prevents crash on game startup.

Still seeing odd timestamps on my MacBook Air M1 (Apple Silicon).  For instance, when a rendering stage is skipped by the game and timestamp diffs should be zero (or near zero), instead I am seeing significant positive numbers.  I do not get this with my Intel Mac running a Radeon 580.  Perhaps the MTLFence issue is not related to the problem I am seeing.

**Comment by billhollings** [comment]
>Still seeing odd timestamps on my MacBook Air M1 (Apple Silicon). For instance, when a rendering stage is skipped by the game and timestamp diffs should be zero (or near zero), instead I am seeing significant positive numbers.

Can you file a new issue here with more details, and a possible way to replicate? And please include information about how significant are the diffs? Are they indicating glitches or hiccups in rendering? Or are they outrageously out of line in a way that appear more like garbage values?
